File: interview/google/hard/LC305.java
public class Solution {

    private int[][] dir = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};

    public List<Integer> numIslands2(int m, int n, int[][] positions) {
        UnionFind2D islands = new UnionFind2D(m, n);
        List<Integer> ans = new ArrayList<>();
        for (int[] position : positions) {
            int x = position[0], y = position[1];
            int p = islands.add(x, y);
            for (int[] d : dir) {
                int q = islands.getID(x + d[0], y + d[1]);
                if (q > 0 && !islands.find(p, q))
                    islands.unite(p, q);
            }
            ans.add(islands.size());
        }
        return ans;
    }
}

class UnionFind2D {
    private int[] id;
    private int[] sz;
    private int m, n, count;

    public UnionFind2D(int m, int n) {
        this.count = 0;
        this.n = n;
        this.m = m;
        this.id = new int[m * n + 1];
        this.sz = new int[m * n + 1];
    }

    public int index(int x, int y) { return x * n + y + 1; }

    public int size() { return this.count; }

    public int getID(int x, int y) {
        if (0 <= x && x < m && 0<= y && y < n)
            return id[index(x, y)];
        return 0;
    }

    public int add(int x, int y) {
        int i = index(x, y);
        id[i] = i; sz[i] = 1;
        ++count;
        return i;
    }

    public boolean find(int p, int q) {
        return root(p) == root(q);
    }

    public void unite(int p, int q) {
        int i = root(p), j = root(q);
        if (sz[i] < sz[j]) { //weighted quick union
            id[i] = j; sz[j] += sz[i];
        } else {
            id[j] = i; sz[i] += sz[j];
        }
        --count;
    }

    private int root(int i) {
        for (;i != id[i]; i = id[i])
            id[i] = id[id[i]]; //path compression
        return i;
    }
}
//Runtime: 20 ms

File: interview/google/hard/LC305.py
class Solution(object):
    
    def numIslands2(self, m, n, positions):
        """
        :type m: int
        :type n: int
        :type positions: List[List[int]]
        :rtype: List[int]
        """
        islandNum = 0
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        union_map = {}
        res = []
        for pos in positions:
            val = pos[0] * n + pos[1]
            union_map[val] = val
            islandNum += 1
            for dir in directions:
                newX = pos[0] + dir[0]
                newY = pos[1] + dir[1]
                newVal = newX * n + newY
                if newX >=0 and newX < m and newY >=0 and newY < n and newVal in union_map:
                    if self.union(val, newVal, union_map):
                        islandNum -= 1
            res.append(islandNum)
        return res
        
    def union(self, m, n, union_map):
        parent_m = self.find(m, union_map)
        parent_n = self.find(n, union_map)
        if parent_m == parent_n:
            return False
        union_map[parent_m] = parent_n
        return True
        
    def find(self, n, union_map):
        if n == union_map[n]:
            return n
        parent = union_map[n]
        # shink the map
        union_map[n] = union_map[parent]
        return self.find(parent, union_map)